src/examples/zkapps/dex/dex-with-actions.ts

```diff
@@ -0,0 +1,367 @@
+      Dex['events'][typeof key]
+    >;
+    this.emitEvent(key, event);
+    return dl;
+  }
+
+  /**
+   * Mint liquidity tokens in exchange for X and Y tokens
+   * @param dx input amount of X tokens
+   * @return output amount of lqXY tokens
+   *
+   * This uses supplyLiquidityBase as the circuit, but for convenience,
+   * the input amount of Y tokens is calculated automatically from the X tokens.
+   * Fails if the liquidity pool is empty, so can't be used for the first deposit.
+   */
+  supplyLiquidity(dx: UInt64): UInt64 {
+    // calculate dy outside circuit
+    let x = Account(this.address, TokenId.derive(this.tokenX)).balance.get();
+    let y = Account(this.address, TokenId.derive(this.tokenY)).balance.get();
+    if (x.value.isConstant() && x.value.isZero().toBoolean()) {
+      throw Error(
+        'Cannot call `supplyLiquidity` when reserves are zero. Use `supplyLiquidityBase`.'
+      );
+    }
+    let dy = dx.mul(y).div(x);
+    return this.supplyLiquidityBase(dx, dy);
+  }
+
+  /**
+   * Burn liquidity tokens to get back X and Y tokens
+   * @param dl input amount of lqXY token
+   *
+   * The transaction needs to be signed by the user's private key.
+   *
+   * NOTE: this does not give back tokens in return for liquidity right away.
+   * to get back the tokens, you have to call {@link DexTokenHolder}.redeemFinalize()
+   * on both token holder contracts, after `redeemInitialize()` has been accepted into a block.
+   *
+   * @emits RedeemAction - action on the Dex account that will make the token holder
+   * contracts pay you tokens when reducing the action.
+   */
+  @method redeemInitialize(dl: UInt64) {
+    this.reducer.dispatch(new RedeemAction({ address: this.sender, dl }));
+    this.token.burn({ address: this.sender, amount: dl });
+    // TODO: preconditioning on the state here ruins concurrent interactions,
+    // there should be another `finalize` DEX method which reduces actions & updates state
+    this.totalSupply.set(this.totalSupply.getAndAssertEquals().sub(dl));
+
+    // emit event
+    let key = 'redeem-liquidity' as const satisfies keyof Dex['events'];
+    let event = { address: this.sender, dl } satisfies InstanceType<
+      Dex['events'][typeof key]
+    >;
+    this.emitEvent(key, event);
+  }
+
+  /**
+   * Helper for `DexTokenHolder.redeemFinalize()` which adds preconditions on
+   * the current action state and token supply
+   */
+  @method assertActionsAndSupply(actionState: Field, totalSupply: UInt64) {
+    this.account.actionState.assertEquals(actionState);
+    this.totalSupply.assertEquals(totalSupply);
+  }
+
+  /**
+   * Swap X tokens for Y tokens
+   * @param dx input amount of X tokens
+   * @return output amount Y tokens
+   *
+   * The transaction needs to be signed by the user's private key.
+   *
+   * Note: this is not a `@method`, since it doesn't do anything beyond
+   * the called methods which requires proof authorization.
```

Comment:
```suggestion
   * the called methods that require proof authorization.
```

Reply:
this is not what I meant - not the methods require proof authorization, but "it doesn't to anything which requires proof authorization". So, the "s" in requires is correct